src/anomalib/metrics/spro.py

```diff
@@ -0,0 +1,158 @@
+"""Implementation of SPRO metric based on TorchMetrics."""
+
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+
+import torch
+from torchmetrics import Metric
+from torchmetrics.utilities.data import dim_zero_cat
+
+logger = logging.getLogger(__name__)
+
+
+class SPRO(Metric):
+    """Saturated Per-Region Overlap (SPRO) Score.
+
+    This metric computes the macro average of the saturated per-region overlap between the
+    predicted anomaly masks and the ground truth masks.
+
+    Args:
+        threshold (float): Threshold used to binarize the predictions.
+            Defaults to ``0.5``.
+        saturation_config (dict): Saturations configuration for each label (pixel value) as the keys.
+            Defaults: ``None`` (which the score is equivalent to PRO metric, but with the 'region' are
+            separated by mask files.
+        kwargs: Additional arguments to the TorchMetrics base class.
+
+    Example:
+        Import the metric from the package:
+
+        >>> import torch
+        >>> from anomalib.metrics import SPRO
+
+        Create random ``preds`` and ``labels`` tensors:
+
+        >>> labels = torch.randint(low=0, high=2, size=(1, 10, 5), dtype=torch.float32)
+        >>> preds = torch.rand_like(labels)
+
+        Compute the SPRO score for labels and preds:
+
+        >>> spro = SPRO(threshold=0.5)
+        >>> spro.update(preds, _, labels)
+        >>> spro.compute()
+        tensor(0.6333)
+
+        .. note::
+            Note that the example above shows random predictions and labels.
+            Therefore, the SPRO score above may not be reproducible.
+
+    """
+
+    targets: list[torch.Tensor]
+    preds: list[torch.Tensor]
+
+    def __init__(self, threshold: float = 0.5, saturation_config: dict | None = None, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.threshold = threshold
+        self.saturation_config = saturation_config
+        self.add_state("preds", default=[], dist_reduce_fx="cat")
+        self.add_state("targets", default=[], dist_reduce_fx="cat")
+
+    def update(self, predictions: torch.Tensor, _: torch.Tensor, masks: torch.Tensor) -> None:
+        """Compute the SPRO score for the current batch.
+
+        Args:
+            predictions (torch.Tensor): Predicted anomaly masks
+            _ (torch.Tensor): Unused argument, but needed for different metrics within the same AnomalibMetricCollection
+            masks (torch.Tensor): Ground truth anomaly masks with non-binary values and, original height and width
+
+        Example:
+            To update the metric state for the current batch, use the ``update`` method:
+
+            >>> spro.update(preds, _, labels)
+        """
+        assert masks is not None
```

Comment:
would it be possible to use explicit error handling?

Reply:
Sure, I will use an explicit error handling here. Or when we decided to remove the `_`, maybe we can safely remove this error check.

Reply:
I have removed the `_`, now this assert is unnecessary and already removed.